### ISSUE TITLE
Add a set of ```trajan.reader.XX``` functions for several standard instruments

### ISSUE THREAD

**Issue by jerabaul29** [OpenDrift/trajan#63]
Some discussion / brainstorming for a possible ```trajan.ingest``` module.

There are a number of instruments we may see again and again, for example, OMBs, SFYs, Sofar spotter etc.

Wondering if it may be useful to have a way to "ingest" data from these instruments, independently of any server / external data scraping tool, by just calling a function on the typical data file generated by these instruments. The rationale is that some users may not be using for example sfy server, and may want to use trajan directly with the instruments they have without the need to go through some middleware.

In the following, taking the example of the OMB, as this is the one I know best.

OMB users will all have access to the ```data.csv``` csv dump provided by Rock7, and this is likely the most "universal" data to start working with, independently of whether they use sfy server or other.

Then, the most general / robust / simple way to make sure OMB users can and have an incitement to use trajan would be to provide them with a python function:

```
trajan.injest.from_omb_csv(path="all.csv")
```

that returns a trajan object.

Do you think that could make sense to have in a new module in trajan?

Some requirements to get over time:

- be dumpable to netcdf
- have all attributes to be CF compliant when dumped to netCDF

**Comment by knutfrode** [comment]
Yes, this makes sense. This is then analogue to `readers` in OpenDrift.

**Comment by gauteh** [comment]
Great 👍 this is probably the currently best starting point. I think it is open whether waves should be on the obs dim, or a separate dimension. Both i think should be supported by Trajan (though in the second case it would be ignored).

**Comment by gauteh** [comment]
https://github.com/OpenDrift/trajan/blob/main/examples/barents.nc.xz

**Comment by gauteh** [comment]
The oo-2023 NC file is an example with wave data.

tir. 11. apr. 2023, 20:21 skrev JR ***@***.***>:

**Comment by jerabaul29** [comment]
Perfect :) .

**Comment by gauteh** [comment]
I think these methods should be available in the top-level module of trajan, to follow the spirit of pandas and xarray (pd.read_csv, pd.read_excel, etc...) but should be implemented in an `import` submodule to keep the code separate:

```
ta.read_omb_csv(..) # (e.g. imported from ta.import.omb.read_omb_csv)
ta.read_spotter(..)
```

**Comment by jerabaul29** [comment]
Trying to get the general structure / architecture in place before implementing. I am not too expert on these extensions over existing package / slightly advanced code structuring, is something like what is pushed there what you think about, or anything I need to change?

https://github.com/jerabaul29/trajan/tree/omb-reader

Should the read_omb_csv function return a ```Traj``` or a ```xr.Dataset``` or other? :)

**Comment by gauteh** [comment]
It should return an `xr.Dataset`, I'll take a look.

**Comment by gauteh** [comment]
The structure looks good, though I think we should go for `import`, `io`, or `readers` as module name.

I think you need to use relative imports.. ? I suggest passing the path argument directly to `pd.read_csv` that way you get the pandas flexibility for free: it could be a file, a path, or even a compressed file (I think). The best is if you can utilize `from_dataframe` in the end (like the code example), that way improvements we make to that in the future will also affect OMB files.

**Comment by jerabaul29** [comment]
I am not too fan of using a python keyword as a module name / part of a module name, so ```import``` feels a bit bad to me, but maybe this is just my personal taste. I can rename to ```readers``` :) .

Thanks for the other tips, will look into this :) .

**Comment by gauteh** [comment]
Yes, that's a good point: trajan.readers is a good choice.

**Comment by jerabaul29** [comment]
@gauteh , regarding the conventions in the final xr / netcdf files: comparing to the OO nc file:

```
(fix_xarray) ~/Downloads> ncdump -h OO-2023.nc 
netcdf OO-2023 {
dimensions:
	trajectory = 7 ;
	obs = 2856 ;
	imu_obs = 692 ;
	frequencies = 55 ;
variables:
	double lon(trajectory, obs) ;
		lon:_FillValue = NaN ;
```

do you think ```gnss_obs``` instead of ```obs``` would be ok? When we start opening for different kinds of obs, wondering if it would be good to specify all of them, not "just the ones that are not gnss"?

**Comment by gauteh** [comment]
In principle I agree, but `obs` is taken from the CF-convention example -- and it is something that trajan would have to detect in some way. It is also possible to put everything into one obs dimension (and just keep lat, lon nan for imu observations), trajan should already handle that.

TL;DR: I don't know :)

**Comment by jerabaul29** [comment]
Ok. My 2 cents, of course I may be wrong / I am very open to changing my opinion / following your lead on what you want trajan to do :) :

- having a single obs dim and using it for both gnss, imu, etc, feels a bit weird / inefficient as there will be plenty of NaNs scattered in the gnss, imu, etc, arrays; while these can be compressed, I think this is potentially confusing for the user and may (?) be harder to parse; so I would vote for keeping 1 obs dimension per observed quantity
- if having several obs dimensions, it feels like either they all should be "qualified", or none of them should be qualified, so obs_XXX where XXX is the kind of the obs for each of them.

CF conventions are very nice but a bit "narrow" feels like sometimes (which is likely a good thing, having narrow conventions makes for easier parsing etc :) ). Wondering if "extending" the convention about having obs_XXX with XXX the name of the obs for files that contain several kinds of obs that are not sampled simultaneously would be a reasonable and nice extension?

**Comment by gauteh** [comment]
I agree, it is better to go with option 2: separate obs-dimensions. However, after some discussion, we think it is better to go with just `obs` for the lat/lon observations. This follows the CF-example, even though it is not perfect. In theory position observations could come from other things than gnss, so if we are to suggest a generic name for position observations for the future CF-convention we should maybe also consider that.

We should also document how we interpret and further narrow / constrict the CF-standard on the trajan documentation site, and eventually suggest to the CF-convention further improvements.

Does this work for now? If it turns out to be a poor choice, I'm open for discussing it again in the future.

**Comment by jerabaul29** [comment]
I lift this convention issue to a new issue, as this is likely something wider than just this omb reader :) .

**Comment by jerabaul29** [comment]
Btw, regarding the reader: do you think it would be ok to pass it a dict of global attrs in addition to the file path, to be used as the global attrs on the final xr / nc result? :)

**Comment by jerabaul29** [comment]
First "running" version (not containing the data yet, but most of code architecture is there) for the reader.omb available at:

https://github.com/jerabaul29/trajan/blob/omb-reader/trajan/readers/omb.py

Let me know if you have comments. I will fill the actual data in the xr result this afternoon :) .

**Comment by jerabaul29** [comment]
(and add some extra attributes etc :) ).

**Comment by gauteh** [comment]
I think that passing attributes as a dictionary is not the best option, see the example here (scroll a bit down) for how to easily do this as a user: https://opendrift.github.io/trajan/autoapi/trajan/index.html#trajan.from_dataframe . By doing that we add another way of adding attributes than what is common in xarray, and deviate from the way this is usually done by xarray. I think sticking as far as possible to xarray when possible makes it easier for users that have learned xarray to know intuitively understand the concepts. Additionally I don't see that it is less work or less complex than doing it using xarray. You need to create the dict anyway, and although you may or may not get rid of a line of code the complexity is pretty much the same.

**Comment by jerabaul29** [comment]
Ok, thanks for the pointers and explanation, that sounds good :) .

**Comment by gauteh** [comment]
https://github.com/jerabaul29/trajan/blob/omb-reader/trajan/readers/omb.py#L57 : Maybe a set and check if required columns are there, but not error if any are missing or in the wrong order.

**Comment by gauteh** [comment]
https://github.com/jerabaul29/trajan/blob/omb-reader/trajan/readers/omb.py#L167: I dont' think you need to worry about this, just use nan's and xarray will worry about that when you create a netcdf file. FillValue etc is netcdf specific as far as I know, and there's lots of logic for it in the netcdf xarray backend.

**Comment by knutfrode** [comment]
Probably not important for the decisions to be made here, but just to mention:

In OpenDrift (as copied from GDAL), it is also possible to open various datasets without actively choosing a reader. Thus you can do
`$ readerinfo <some file>`  and OpenDrift will test various readers until one of them are able to read the file without errors.
Similarly you can do `gdal.open(<some file>)` without actively choosing a reader (or `driver` in their terminology). In both cases you can be specific if desired.
Yes, its for laziness, but the saved energy/time can be used for other useful things :)

**Comment by gauteh** [comment]
Yes, good point. That would require a `detect` for each reader to try and detect whether it can be opened with that reader, starting from the most specific.

**Comment by jerabaul29** [comment]
Thanks for the feedback :) 

> https://github.com/jerabaul29/trajan/blob/omb-reader/trajan/readers/omb.py#L57 : Maybe a set and check if required columns are there, but not error if any are missing or in the wrong order.

I think I have seen these exact columns in this exact order for many years from Rock7; if ok I would like to keep this as is - likely if these exact columns in this exact order are not here as tested, then this is either a corrupted file, or there has been a major Rock7 API updates, or something like this - I think that in this case it would be nice to be notified and have a look to check nothing is broken silently :) .

> https://github.com/jerabaul29/trajan/blob/omb-reader/trajan/readers/omb.py#L116 : use a specific exception so that unrelated errors are caught.

Agree this is ugly. The issue I see is that I may have plenty of weird errors coming from the debugger (in theory; in practice I have seen relatively few decoder failures), so not sure which type(s) of exceptions I should specifically catch here... Do you think an ok fix for now would be to only try catch the decoder itself (so this will try catch only the underlying decoder), and put the other logics that may be wrongly caught outside? I just pushed this here, do you think this is acceptable compromise?

https://github.com/jerabaul29/trajan/blob/0c1ac9ceb0c439230fbaf8a61b9b46123dcbe178/trajan/readers/omb.py#L84-L90

> https://github.com/jerabaul29/trajan/blob/omb-reader/trajan/readers/omb.py#L167: I dont' think you need to worry about this, just use nan's and xarray will worry about that when you create a netcdf file. FillValue etc is netcdf specific as far as I know, and there's lots of logic for it in the netcdf xarray backend.

Actually, I tried to set NaN at first, but then xarray seems to always "degrade" my arrays to being float-typed (if I understand correctly, int64 for example does not have any representation for +-Inf and NaN in the libraries used, or at least some of them). I do not like having float-typed for an exact quantity such as "seconds since", which is why I do it so. If you know of another way that is better let me know, I am interested to hear about it :) .

**Comment by gauteh** [comment]
> > https://github.com/jerabaul29/trajan/blob/omb-reader/trajan/readers/omb.py#L57 : Maybe a set and check if required columns are there, but not error if any are missing or in the wrong order.
> 
> I think I have seen these exact columns in this exact order for many years from Rock7; if ok I would like to keep this as is - likely if these exact columns in this exact order are not here as tested, then this is either a corrupted file, or there has been a major Rock7 API updates, or something like this - I think that in this case it would be nice to be notified and have a look to check nothing is broken silently :) .

Sounds good.

> 
> > https://github.com/jerabaul29/trajan/blob/omb-reader/trajan/readers/omb.py#L116 : use a specific exception so that unrelated errors are caught.
> 
> Agree this is ugly. The issue I see is that I may have plenty of weird errors coming from the debugger (in theory; in practice I have seen relatively few decoder failures), so not sure which type(s) of exceptions I should specifically catch here... Do you think an ok fix for now would be to only try catch the decoder itself (so this will try catch only the underlying decoder), and put the other logics that may be wrongly caught outside? I just pushed this here, do you think this is acceptable compromise?

I always prefer errors to happen rather than ignoring and proceeding, so if the chance of error is low maybe it is better to not have try/except at all. If not, it's ok, but you will ignore errors like trying to import something that is not installed etc.

> 
> https://github.com/jerabaul29/trajan/blob/0c1ac9ceb0c439230fbaf8a61b9b46123dcbe178/trajan/readers/omb.py#L84-L90
> 
> > https://github.com/jerabaul29/trajan/blob/omb-reader/trajan/readers/omb.py#L167: I dont' think you need to worry about this, just use nan's and xarray will worry about that when you create a netcdf file. FillValue etc is netcdf specific as far as I know, and there's lots of logic for it in the netcdf xarray backend.
> 
> Actually, I tried to set NaN at first, but then xarray seems to always "degrade" my arrays to being float-typed (if I understand correctly, int64 for example does not have any representation for +-Inf and NaN in the libraries used, or at least some of them). I do not like having float-typed for an exact quantity such as "seconds since", which is why I do it so. If you know of another way that is better let me know, I am interested to hear about it :) .

Ah right, yes, nan's only exist for floats. I guess you need to use masked arrays to get the same result with int's, just be sure that it also works if someone saves this to a postgres server or a zarr file or whatever and not netcdf. Does it work correctly without first saving to netcdf?

**Comment by gauteh** [comment]
> > Ah right, yes, nan's only exist for floats. I guess you need to use masked arrays to get the same result with int's, just be sure that it also works if someone saves this to a postgres server or a zarr file or whatever and not netcdf. Does it work correctly without first saving to netcdf?
> 
> At least it works in the playing around I did :) . Maybe we can keep it as is for now, and if there are any complaints, come back to this? :)

Ok. I see that this is the time steps. I would put in a pd.to_datetime(..) array and let xarray handle the rest. The 'gregorian calender.. since' stuff is also handled automatically by xarray then depending on the datetime array. Now you risk specifying attributes that conflict or are over-written anyway by xarray.

**Comment by gauteh** [comment]
Ok, sounds good. I'm not sure it is better, but I prefer to let xarray and pandas handle that.

**Comment by jerabaul29** [comment]
The issues I have found is that different packages have different opinions: there are of course "standard" datetime.datetime, then numpy tries to push on us datetime64 with different [time_units], pandas has its own stuff, so does matplotlib, etc. My feeling so far (but may be wrong) is that the only "universal" thing I always easily manage to handle across these packages is some form of int64 since epoch, at least they all can understand this :) .

**Comment by knutfrode** [comment]
I have felt for a long time that the handling of times in Python with datetime, np.datetime64 and cftime etc has become a real mess.
The strange name "datetime64" is in itself an indicator of ad hoc solutions.

**Comment by jerabaul29** [comment]
I just pushed a version that seems ok on my end (tested on a few of the deployments around Svalbard). Feel free to comment, and let me know when you feel I should open a pull request :) .
